sockeye/image_captioning/data_io.py

```diff
@@ -0,0 +1,419 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License
+# is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+Implements data iterators and I/O related functions for image-to-sequence
+models.
+"""
+import logging
+import time
+from typing import Any, Dict, Iterable, List, Optional, Tuple
+
+import mxnet as mx
+import numpy as np
+
+from .utils import load_features, load_feature, load_preprocess_images
+from .. import constants as C
+from .. import vocab
+from ..data_io import ParallelDataSet, RawParallelDatasetLoader, \
+    BucketBatchSize, FileListReader, SequenceReader, DataConfig, DataInfo, \
+    ParallelSampleIter
+from ..data_io import get_target_bucket, get_data_statistics, \
+    define_empty_source_parallel_buckets, define_bucket_batch_sizes
+
+logger = logging.getLogger(__name__)
+
+try:  # Try to import pillow
+    import matplotlib.pyplot as plt  # pylint: disable=import-error
```

Comment:
can we move this import into `visualize_batch()`? 
When I check out your branch and run pytest I am getting the following error:
```
test/unit/image_captioning/test_data_io.py:23: in <module>
    import sockeye.image_captioning.data_io as data_io
sockeye/image_captioning/data_io.py:37: in <module>
    import matplotlib.pyplot as plt  # pylint: disable=import-error
../miniconda3/lib/python3.6/site-packages/matplotlib/pyplot.py:115: in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
../miniconda3/lib/python3.6/site-packages/matplotlib/backends/__init__.py:32: in pylab_setup
    globals(),locals(),[backend_name],0)
../miniconda3/lib/python3.6/site-packages/matplotlib/backends/backend_macosx.py:19: in <module>
    from matplotlib.backends import _macosx
E   RuntimeError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are using (Ana)Conda please install python.app and replace the use of 'python' with 'pythonw'. See 'Working with Matplotlib on OSX' in the Matplotlib FAQ for more information.
```

I would know how to fix it but other may not and we should make sure that the tests run, regardless whether someone has installed matplotlib or not

Reply:
We can move it to `visualize_batch` although suboptimal.

Ok I will add:
```
import matplotlib
matplotlib.use("Agg")
```

Reply:
Its only used for `visualize_batch()` though right? Why would it be suboptimal?
If you want to avoid the iterative import you could condition on a global variable in this namespace.

Reply:
Makes sense.